tests/test_bootstrap.py

```diff
@@ -0,0 +1,556 @@
+    this module.
+
+    .. warning:: This is supposed to be used as a base class, so please, don't
+                 use this directly.
+    """
+
+    @property
+    def bootstrap_name(self):
+        """Subclasses must have property 'bootstrap_name'. It should be the
+        name of the bootstrap to test"""
+        raise NotImplementedError("Not implemented in GenericBootstrapTest")
+
+    @mock.patch("pythonforandroid.bootstraps.service_only.open", create=True)
+    @mock.patch("pythonforandroid.bootstraps.webview.open", create=True)
+    @mock.patch("pythonforandroid.bootstraps.sdl2.open", create=True)
+    @mock.patch("pythonforandroid.distribution.open", create=True)
+    @mock.patch(
+        "pythonforandroid.python.GuestPythonRecipe.create_python_bundle"
+    )
+    @mock.patch("pythonforandroid.bootstrap.Bootstrap.strip_libraries")
+    @mock.patch("pythonforandroid.util.exists")
+    @mock.patch("pythonforandroid.util.chdir")
+    @mock.patch("pythonforandroid.bootstrap.listdir")
+    @mock.patch("pythonforandroid.bootstrap.sh.rm")
+    @mock.patch("pythonforandroid.bootstrap.sh.cp")
+    def test_run_distribute(
+        self,
+        mock_sh_cp,
+        mock_sh_rm,
+        mock_listdir,
+        mock_chdir,
+        mock_ensure_dir,
+        mock_strip_libraries,
+        mock_create_python_bundle,
+        mock_open_dist_files,
+        mock_open_sdl2_files,
+        mock_open_webview_files,
+        mock_open_service_only_files,
+    ):
+        """
+        A test for any overwritten method of
+        `~pythonforandroid.bootstrap.Bootstrap.run_distribute`. Here we mock
+        any file/dir operation that it could slow down our tests, and there is
+        a lot to mock, because the `run_distribute` method it should take care
+        of prepare all compiled files to generate the final `apk`. The targets
+        of this test will be:
+
+            - :meth:`~pythonforandroid.bootstraps.sdl2.BootstrapSdl2
+              .run_distribute`
+            - :meth:`~pythonforandroid.bootstraps.service_only
+              .ServiceOnlyBootstrap.run_distribute`
+            - :meth:`~pythonforandroid.bootstraps.webview.WebViewBootstrap
+               .run_distribute`
+            - :meth:`~pythonforandroid.bootstraps.empty.EmptyBootstrap.
+              run_distribute`
+
+        Here we will tests all those methods that are specific for each class.
+        """
+        # prepare bootstrap and distribution
+        bs = Bootstrap().get_bootstrap(self.bootstrap_name, self.ctx)
+        bs.build_dir = bs.get_build_dir()
+        self.setUp_distribution_with_bootstrap(bs)
+
+        self.ctx.hostpython = "/some/fake/hostpython3"
+        self.ctx.python_recipe = Recipe.get_recipe("python3", self.ctx)
+        self.ctx.python_modules = ["requests"]
+        self.ctx.archs = [ArchARMv7_a(self.ctx)]
+
+        mock_open_dist_files.side_effect = [
+            mock.mock_open(read_data="target=android-21").return_value
+        ]
```

Comment:
Sorry this is still confusing me. What are these for again? :grimacing:

Reply:
Yes, this is wrong  :sweat_smile:, I wanted to check that we actually `wrote` the expected values and this is not the way to go, I will change that

**Thanks!!**